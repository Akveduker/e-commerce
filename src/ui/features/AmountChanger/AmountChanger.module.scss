@import "../../../styles/_vars.scss";
.container {
    display: flex;
    align-items: center;
    background-color: $C1H;
    border-radius: $borderRadius;
    border: 1px $C1D solid;
    padding: 12.5px 14px;
    font-size: 14px;
    max-width: 111px;
    &__small {
        padding: 6.5px 6px;
    }
}
.button {
    border: none;
    font-size: inherit;
    cursor: pointer;
}
.input {
    text-align: center;
    width: 100%;
    border: none;
    background-color: $C1H;
    font-size: inherit;
}
