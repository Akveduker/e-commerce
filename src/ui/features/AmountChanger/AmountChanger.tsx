import React, { FC, memo, useEffect, useLayoutEffect, useState } from 'react';
import { onlyNumbersInput } from '../../../utils/filters/onlyNumbersInput/onlyNumbersInput';
import s from './AmountChanger.module.scss'
interface AmountChangerProps {
    isSmall?: boolean;
    setAmount: React.Dispatch<React.SetStateAction<number>>;
    amount: number;
}
const AmountChanger: FC<AmountChangerProps> = ({ isSmall, setAmount, amount }) => {
    const className = isSmall ? `${s['container']} ${s['container__small']}` : s['container']
    const [value, setValue] = useState(amount)
    useLayoutEffect(() => {
        setValue(amount)
    }, [amount])
    useEffect(() => {
        if (value > 0 && value !== amount) setAmount(value)
    }, [value])
    const inputOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const inputValue = e.target.value
        if (onlyNumbersInput(inputValue)) {
            +inputValue <= 0 ? setValue(0) : setValue(+inputValue)
        }
    }
    const inputOnBlur = () => {
        +value <= 1 ? setValue(1) : setValue(+value)
    }
    const buttonChangeMinus = () => {
        value - 1 <= 1 ? setValue(1) : setValue(value - 1)
    }
    return (
        <div className={className}>
            <button onClick={() => setValue(value + 1)} className={s['button']} data-testid={'plus'}>
                +
            </button>
            <input type="text" value={value} className={s['input']} onChange={inputOnChange} onBlur={inputOnBlur} data-testid={'input'} />
            <button onClick={buttonChangeMinus} className={s['button']} data-testid={'minus'}>
                -
            </button>
        </div>
    );
};

export default memo(AmountChanger);