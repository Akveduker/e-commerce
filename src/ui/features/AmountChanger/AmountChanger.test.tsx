import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import AmountChanger from "./AmountChanger"

describe('AmountChanger', () => {
    const amount = 10
    const amountChanger = jest.fn
    test('Снапшот', () => {
        const { container } = render(
            <AmountChanger
                amount={amount}
                setAmount={amountChanger}
            />)
        expect(container).toMatchSnapshot()
    })
    describe('Плюс', () => {
        test('прибавить', () => {
            render(
                <AmountChanger
                    amount={amount}
                    setAmount={amountChanger}
                />)
            const plus = screen.getByTestId('plus')
            const input = screen.getByTestId('input')
            userEvent.click(plus)
            expect(input).toHaveValue('11')
        })
    })
    describe('Минус', () => {
        test('Отнять', () => {
            render(
                <AmountChanger
                    amount={amount}
                    setAmount={amountChanger}
                />)
            const plus = screen.getByTestId('minus')
            const input = screen.getByTestId('input')
            userEvent.click(plus)
            expect(input).toHaveValue('9')
        })
        test('Отнять меньше 1', () => {
            const amount = 0;
            render(
                <AmountChanger
                    amount={amount}
                    setAmount={amountChanger}
                />)
            const plus = screen.getByTestId('minus')
            const input = screen.getByTestId('input')
            userEvent.click(plus)
            expect(input).toHaveValue('1')
        })
    })

})