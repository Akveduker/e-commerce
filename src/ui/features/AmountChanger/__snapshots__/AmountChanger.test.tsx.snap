// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountChanger Снапшот 1`] = `
<div>
  <div
    class="container"
  >
    <button
      class="button"
      data-testid="plus"
    >
      +
    </button>
    <input
      class="input"
      data-testid="input"
      type="text"
      value="10"
    />
    <button
      class="button"
      data-testid="minus"
    >
      -
    </button>
  </div>
</div>
`;
